### PR TITLE
[core] Reduce symbol search only to when autoloading is enabled.

### PR DESCRIPTION
**Pull request by vgvassilev**
The llvm9 JIT issued callbacks when a symbol was missing and we reacted on it by loading the relevant library. In root-project/root@9b2041e3 we have kept the logic but now the JIT started querying more often even for symbols which are okay to be missing. In turn that leads to scanning all libraries causing performance issues.

This patch tries to limit this functionality only in contexts where automatic loading is allowed. In addition when computing the offsets of a constant variable declaration we compute the initializers instead of searching in the shared objects.

### CODE REVIEW

**Comment by pcanal** [comment]
This patch seems to be a good performance enhancement (skipping some work that will not be used) but I am confused on how it helps with the increase in autoloading as I would have expected that in the case where this patch is active (i.e. autoloading is disabled) the net effect would have been the same (i.e. autoloading not activated).  What am I missing?

i.e. this patch looks good to me as-is but does not seem to fully address the problem.

**Comment by vepadulano** [comment]
Reporting some numbers from my machine with this patch, building ROOT with the following cmake invocation:
```
cmake -DCMAKE_BUILD_TYPE=Release -S rootsrc -B rootbuild/BUILD_NAME -DCMAKE_INSTALL_DIR=rootinstall/INSTALL_NAME
```

TLDR: The situation w.r.t. #14277 is seemingly unchanged

## Baseline: [master@fbd6abb](https://github.com/root-project/root/commit/fbd6abb371e8611a16648983eedf095e425bebd4)

### Import time
Taken after a few runs, hot caches

```
$: time python -c "import ROOT"

real	0m0.791s
user	0m0.592s
sys	0m0.203s
```

### strace file stats

```
strace -e file python -c 'import ROOT' 2>&1 | awk -v FS='"' '{ print $1 }' | sort | uniq -c
    206 access(
      2 chdir(
      1 execve(
      1 +++ exited with 0 +++
    181 getcwd(
      2 newfstatat(0, 
      2 newfstatat(1, 
      2 newfstatat(2, 
    532 newfstatat(3, 
     29 newfstatat(4, 
  10123 newfstatat(AT_FDCWD, 
    546 openat(AT_FDCWD, 
    843 readlink(
      1 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=130118, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
      1 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=130120, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
      1 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=130124, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
```

### openat stats with the two symbols related to #14277 

```
strace -z -f -o openat.log -e trace=open,openat python3 -c 'import ROOT;print(ROOT.gErrorIgnoreLevel);print(ROOT.kError)'
-1
3000
grep openat openat.log | sed 's|.*, "||;s|".*||' | sort | uniq | wc -l
338
```

## After this patch

### Import time
Taken after a few runs, hot caches

```
time python -c "import ROOT"

real	0m0.792s
user	0m0.600s
sys	0m0.197s
```

### strace file stats

```
strace -e file python -c 'import ROOT' 2>&1 | awk -v FS='"' '{ print $1 }' | sort | uniq -c
    206 access(
      2 chdir(
      1 execve(
      1 +++ exited with 0 +++
    181 getcwd(
      2 newfstatat(0, 
      2 newfstatat(1, 
      2 newfstatat(2, 
    532 newfstatat(3, 
     29 newfstatat(4, 
  10123 newfstatat(AT_FDCWD, 
    546 openat(AT_FDCWD, 
    843 readlink(
      1 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=130864, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
      1 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=130866, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
      1 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=130870, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
```

### openat stats with the two symbols related to #14277 

```
strace -z -f -o openat.log -e trace=open,openat python3 -c 'import ROOT;print(ROOT.gErrorIgnoreLevel);print(ROOT.kError)'
-1
3000
grep openat openat.log | sed 's|.*, "||;s|".*||' | sort | uniq | wc -l
338
```

**Comment by vgvassilev** [comment]
Don’t spend time on this patch just yet. I have more local changes that I am going to upload soon-ish.

**Comment by vgvassilev** [comment]
@vepadulano, can you run your tests again. With the current changes we support `constexpr` and it should be fast as we do not even try to look for the symbols but we evaluate the initializer instead.

**Comment by vepadulano** [comment]
# Updates 

## Baseline: master before #14261 and #14276 

```
time python -c 'import ROOT;print(ROOT.gErrorIgnoreLevel);print(ROOT.kError)'
-1
3000

real	0m1.198s
user	0m0.706s
sys	0m0.475s
```

```
nm -C $ROOTSYS/lib/*.so | grep TString::kNPOS | wc -l
0
nm -C $ROOTSYS/lib/*.so | grep kError | wc -l
27
```

```
strace -e file python -c 'import ROOT;print(ROOT.gErrorIgnoreLevel);print(ROOT.kError)' 2>&1 | awk -v FS='"' '{ print $1 }' | sort | uniq -c
      1 -1
      1 3000
    219 access(
      2 chdir(
      1 execve(
      1 +++ exited with 0 +++
    184 getcwd(
      2 newfstatat(0, 
      2 newfstatat(1, 
      2 newfstatat(2, 
  10455 newfstatat(3, 
   3997 newfstatat(4, 
  31461 newfstatat(AT_FDCWD, 
  19903 openat(AT_FDCWD, 
  28769 readlink(
      1 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=11626, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
      1 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=11628, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
      1 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=11632, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
```
```
strace -z -f -o openat.log -e trace=open,openat python3 -c 'import ROOT;print(ROOT.gErrorIgnoreLevel);print(ROOT.kError)'
grep openat openat.log | wc -l
19759
```

## Master after the two hotfixes

```
time python -c "import ROOT;print(ROOT.gErrorIgnoreLevel);print(ROOT.kError)"
-1
3000

real	0m0.807s
user	0m0.620s
sys	0m0.183s
```

```
nm -C $ROOTSYS/lib/*.so | grep TString::kNPOS | wc -l
1
nm -C $ROOTSYS/lib/*.so | grep kError | wc -l
34
```

```
strace -e file python -c 'import ROOT;print(ROOT.gErrorIgnoreLevel);print(ROOT.kError)' 2>&1 | awk -v FS='"' '{ print $1 }' | sort | uniq -c
      1 -1
      1 3000
    219 access(
      2 chdir(
      1 execve(
      1 +++ exited with 0 +++
    184 getcwd(
      2 newfstatat(0, 
      2 newfstatat(1, 
      2 newfstatat(2, 
    532 newfstatat(3, 
     29 newfstatat(4, 
  10141 newfstatat(AT_FDCWD, 
    546 openat(AT_FDCWD, 
    843 readlink(
      1 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=12038, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
      1 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=12040, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
      1 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=12044, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
```

```
 strace -z -f -o openat.log -e trace=open,openat python3 -c 'import ROOT;print(ROOT.gErrorIgnoreLevel);print(ROOT.kError)'
-1
3000
grep openat openat.log | wc -l
402
```

## This PR

```
time python -c "import ROOT;print(ROOT.gErrorIgnoreLevel);print(ROOT.kError)"
-1
3000

real	0m0.805s
user	0m0.612s
sys	0m0.191s
```

```
nm -C $ROOTSYS/lib/*.so | grep TString::kNPOS | wc -l
1
nm -C $ROOTSYS/lib/*.so | grep kError | wc -l
27
```

```
strace -e file python -c 'import ROOT;print(ROOT.gErrorIgnoreLevel);print(ROOT.kError)' 2>&1 | awk -v FS='"' '{ print $1 }' | sort | uniq -c
      1 -1
      1 3000
    219 access(
      2 chdir(
      1 execve(
      1 +++ exited with 0 +++
    184 getcwd(
      2 newfstatat(0, 
      2 newfstatat(1, 
      2 newfstatat(2, 
    532 newfstatat(3, 
     29 newfstatat(4, 
  10141 newfstatat(AT_FDCWD, 
    546 openat(AT_FDCWD, 
    843 readlink(
      1 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=12243, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
      1 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=12245, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
      1 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=12249, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
```

```
strace -z -f -o openat.log -e trace=open,openat python3 -c 'import ROOT;print(ROOT.gErrorIgnoreLevel);print(ROOT.kError)'
-1
3000
grep openat openat.log | wc -l
402
```

**Comment by vepadulano** [comment]
@vgvassilev It seems that the changes in this PR provide the same performance as the current master (after the two hotfixes), while supporting leaving the `k*` constants in `TError.h` as `constexpr` so without producing a symbol in `libCore.so`. For `TString::kNPOS` I guess we want to leave the commit there since it was anyway a real fix to a bug.

Another problem which is still present in both the current master and even after this PR is that the number of file operations is still quite high, i.e.
```
    532 newfstatat(3, 
     29 newfstatat(4, 
  10141 newfstatat(AT_FDCWD, 
    546 openat(AT_FDCWD, 
    843 readlink(
```

which are exactly the same whether current master or this patch. Consider that many of those `newfstatat` calls are spurious and I'm trying to understand where they come from. Take the following as an example
```
newfstatat(AT_FDCWD, "/home/vpadulan/programs/rootproject/rootbuild/reduce-symbol-search-release/gui/sessionviewer/input_line_11", 0x7ffde442f050, 0) = -1 ENOENT (No such file or directory)
```

Note the presence of the `rootbuild` directory, which is the path where I store the ROOT build artifacts, even though I am using ROOT from the install directory. I thought it was something in the machinery of `(T)cling` but so far my tests haven't brought me to any conclusion. Another suspicion is that the path to the build directory is hardcoded somehow in the `.pcm` or in the AST and doesn't get converted to the install directory.

**Comment by vgvassilev** [comment]
> @vgvassilev It seems that the changes in this PR provide the same performance as the current master (after the two hotfixes), while supporting leaving the `k*` constants in `TError.h` as `constexpr` so without producing a symbol in `libCore.so`. 

Thanks for doing this. That sounds good

> For `TString::kNPOS` I guess we want to leave the commit there since it was anyway a real fix to a bug.

I agree, we should also land this PR since it supports now constexpr constructs which can be anywhere in external codes.

> Consider that many of those newfstatat calls are spurious and I'm trying to understand where they come from. 

These are likely garbage that's put by rootcling into the pcm files. We can verify if that theory is right if we make a build with `-Druntime_cxxmodules=Off` and see if we still get them at such high numbers...

**Comment by vepadulano** [comment]
@vgvassilev indeed you were right, the difference is huge! I opened an issue to keep track of this https://github.com/root-project/root/issues/14296 . If we understand how to fix the paths in the PCMs at install time there is an opportunity for a very substantial performance improvement!

**Comment by vgvassilev** [comment]
We understand how to fix them but that require reworking of rootcling to a good extent. When we have Interpreter::Declare (or similar) in rootcling we create a virtual file (input_line_N) and it gets stored in the pcms (also in the pch but we have 1 pch vs many pcms). When we load a pcm clang tries to stat the files to see if the sources are around.

**Comment by pcanal** [comment]
> For TString::kNPOS I guess we want to leave the commit there since it was anyway a real fix to a bug.

If it can be turned into a `constexpr` that would probably be better.

**Comment by vgvassilev** [comment]
@phsft-bot build just on windows10/default

**Comment by vgvassilev** [comment]
@phsft-bot build!

**Comment by vepadulano** [comment]
> If it can be turned into a constexpr that would probably be better.

Here are the numbers, after this PR + the diff
```
diff --git a/core/base/inc/TString.h b/core/base/inc/TString.h
index 7710169876..8181dc2563 100644
--- a/core/base/inc/TString.h
+++ b/core/base/inc/TString.h
@@ -277,7 +277,7 @@ private:
 public:
    enum EStripType   { kLeading = 0x1, kTrailing = 0x2, kBoth = 0x3 };
    enum ECaseCompare { kExact, kIgnoreCase };
-   static const Ssiz_t kNPOS = ::kNPOS;
+   static constexpr Ssiz_t kNPOS = ::kNPOS;
 
    TString();                           // Null string
    explicit TString(Ssiz_t ic);         // Suggested capacity
diff --git a/core/base/src/TString.cxx b/core/base/src/TString.cxx
index 4087cdbb52..f6efde6a0a 100644
--- a/core/base/src/TString.cxx
+++ b/core/base/src/TString.cxx
@@ -54,12 +54,6 @@ as a TString, construct a TString from it, eg:
 #include "TVirtualMutex.h"
 #include "ThreadLocalStorage.h"
 
-// Definition of the TString static data member. Declaration (even with
-// initialization) in the class body *is not* definition according to C++
-// standard. The definition must be explicitly done in one TU for ODR use. See
-// https://en.cppreference.com/w/cpp/language/definition
-const Ssiz_t TString::kNPOS;
```

```
time python -c "import ROOT;print(ROOT.gErrorIgnoreLevel);print(ROOT.kError)"
-1
3000

real	0m0.805s
user	0m0.604s
sys	0m0.198s
```
```
nm -C $ROOTSYS/lib/*.so | grep TString::kNPOS | wc -l
0
```
strace -z -f -o openat.log -e trace=open,openat python3 -c 'import ROOT;print(ROOT.gErrorIgnoreLevel);print(ROOT.kError)'
-1
3000
grep openat openat.log | wc -l
402
```

So indeed, when this PR is merged we can make `TString::kNPOS` `constexpr` and remove also that symbol from `libCore.so` :+1:

**Comment by pcanal** [comment]
Great.   Several `const` at the end of `core/foundation/inc/RtypesCore.h` should also be updated in a similar manner (including `::kNPOS`)

**Comment by vgvassilev** [comment]
@smuzaffar, can we run verify that this PR works with cmssw?

**Comment by smuzaffar** [comment]
sure, started the tests at https://github.com/cms-sw/root/pull/193

**Comment by dpiparo** [comment]
Hi @smuzaffar , looking at the CMSSW tests, it seems this patch worked for CMS: could you confirm that?

**Review by dpiparo — APPROVED** [review]
Thanks @vgvassilev for this fix, delivered so quickly. I propose to merge the code if other reviewers approve the PR and if 2 conditions are met:
1. We get the confirmation from CMS that this is all what they need. 
2. We understand why `roottest-python-basic-basic` fails on fedora38 and, if possible, fix the problem.

**Comment by dpiparo** [comment]
(btw, since we are looking for ways to speedup ROOT startup, I think this could also be an avenue https://github.com/root-project/root/issues/7123)

File: core/metacling/src/TCling.cxx [hunk 1]
```diff
@@ -4726,6 +4726,9 @@ TInterpreter::DeclId_t TCling::GetDataMember(ClassInfo_t *opaque_cl, const char
    DeclId_t d;
    TClingClassInfo *cl = (TClingClassInfo*)opaque_cl;
 
+   // Could trigger deserialization of decls.
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply core/metacling/src/TCling.cxx:4729]: Because the tests make `GetDataMember` now deserialize.

**Comment by smuzaffar** [comment]
> Hi @smuzaffar , looking at the CMSSW tests, it seems this patch worked for CMS: could you confirm that?

Yes @dpiparo , this worked for CMS too. Local tests also show that with this update CMSSW/ROOT symbols mentioned in https://github.com/cms-sw/cmssw/issues/43077#issuecomment-1867640888 also does not trigger loading of all shared libs.

File: core/metacling/src/TClingCallbacks.cxx [hunk 2]
```diff
@@ -158,13 +164,18 @@ class AutoloadLibraryMU : public llvm::orc::MaterializationUnit {
 };
 
 class AutoloadLibraryGenerator : public llvm::orc::DefinitionGenerator {
+   const TClingCallbacks &CB;
```
Comment (hahnjo) [inline core/metacling/src/TClingCallbacks.cxx:167]: Can you please put the data member next to the others, and rename it as `fCallbacks` to be consistent?

File: core/metacling/src/TClingDataMemberInfo.cxx [hunk 3]
```diff
@@ -352,10 +352,9 @@ Longptr_t TClingDataMemberInfo::Offset()
       //   static constexpr Long64_t something = std::numeric_limits<Long64_t>::max();
       cling::Interpreter::PushTransactionRAII RAII(fInterp);
 
-      if (Longptr_t addr = reinterpret_cast<Longptr_t>(fInterp->getAddressOfGlobal(GlobalDecl(VD))))
-         return addr;
-      auto evalStmt = VD->ensureEvaluatedStmt();
-      if (evalStmt && evalStmt->Value) {
+      // We can't reassign constexpr or const variables. We can compute the
+      // initializer.
+      if (VD->hasInit() && (VD->isConstexpr() || VD->getType().isConstQualified())) {
```
Comment (hahnjo) [inline core/metacling/src/TClingDataMemberInfo.cxx:357]: I think this would hugely benefit from being its own commit, for logical reasons and also for future `git-bisect` runs.

File: interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp [hunk 4]
```diff
@@ -240,9 +240,20 @@ void IncrementalExecutor::replaceSymbol(const char* Name, void* Addr) const {
...
+      *fromJIT = false;
+    return Addr;
+  }
+
+  // Try iterating over the registered llvm libraries.
+  if (void* Addr = llvm::sys::DynamicLibrary::SearchForAddressOfSymbol(symbolName.str())) {
+    if (fromJIT)
+      *fromJIT = false;
+    return Addr;
+  }
```
Comment (hahnjo) [inline interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp:256]: This looks like a somewhat separate performance optimization that isn't explained in the commit message. Should this be its own commit?

File: core/foundation/inc/TError.h [hunk 5]
```diff
@@ -39,14 +39,15 @@
...
-R__EXTERN const Int_t kSysError;
-R__EXTERN const Int_t kFatal;
+constexpr Int_t kUnset    =  -1;
+constexpr Int_t kPrint    =   0;
+constexpr Int_t kInfo     =   1000;
+constexpr Int_t kWarning  =   2000;
+constexpr Int_t kError    =   3000;
+constexpr Int_t kBreak    =   4000;
+constexpr Int_t kSysError =   5000;
+constexpr Int_t kFatal    =   6000;
```
Comment (hahnjo) [inline core/foundation/inc/TError.h:49]: This revert must be logically *after* the issue is fixed in Cling - please swap the two commits!

**Comment by hahnjo** [comment]
> We understand why `roottest-python-basic-basic` fails on fedora38 and, if possible, fix the problem.

I believe this is not related to this PR, some of the Python tests fail from time to time on Fedora 38 and Fedora 39. I suspect there are still GC-related issues with Python 3.11 and 3.12 that were never fully understood...

File: interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp (hunk 4 above)
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp]: I've removed that from the PR for now - we need to rethink the `fromJIT` thingy anyways...

File: core/metacling/src/TClingDataMemberInfo.cxx (hunk 3 above)
Comment (vgvassilev) [reply core/metacling/src/TClingDataMemberInfo.cxx:357]: Agreed, that's the entire commit with the test included.

File: core/metacling/src/TClingCallbacks.cxx [hunk 6]
```diff
@@ -84,16 +84,22 @@ extern "C" {
...
 
    StringRef getName() const override { return "<Symbols from Autoloaded Library>"; }
 
    void materialize(std::unique_ptr<llvm::orc::MaterializationResponsibility> R) override
    {
+      if (!CB.IsAutoLoadingEnabled()) {
+         R->failMaterialization();
+         return;
+      }
+
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply core/metacling/src/TClingCallbacks.cxx]: I had a quick chat with Lang and IIUC returning `failMaterialization` is fine here.

File: core/metacling/src/TClingDataMemberInfo.cxx (hunk 3 above)
Comment (hahnjo) [reply core/metacling/src/TClingDataMemberInfo.cxx:357]: No, it's still mudged together with the changes in `AutoloadLibraryMU`. If I understand correctly, they are two separate changes, no?

File: core/metacling/src/TClingCallbacks.cxx (hunk 6 above)
[the remark this answers is not in the document]
Comment (hahnjo) [reply core/metacling/src/TClingCallbacks.cxx]: I'm not doubting that, but I'm saying that it's not needed: If there are not MUs created, they don't need to check for `IsAutoLoadingEnabled` a second time...

File: core/metacling/src/TClingDataMemberInfo.cxx (hunk 3 above)
Comment (vgvassilev) [reply core/metacling/src/TClingDataMemberInfo.cxx:357]: No, the changes in the `AutoloadLibraryMU` enable the tests in the unittests, otherwise there is not really a clear way how to test that change there.

File: core/metacling/src/TClingCallbacks.cxx [hunk 7]
```diff
@@ -84,16 +84,22 @@ extern "C" {
 }
 
 class AutoloadLibraryMU : public llvm::orc::MaterializationUnit {
+   const TClingCallbacks &CB;
```
[the remark this answers is not in the document]
Comment (hahnjo) [reply core/metacling/src/TClingCallbacks.cxx]: To reduce the diff, it would have been easier to move the new member next to the existing ones. IIRC I'm following upstream style here by having the members last in each `class`...

File: core/metacling/src/TClingDataMemberInfo.cxx (hunk 3 above)
Comment (hahnjo) [reply core/metacling/src/TClingDataMemberInfo.cxx:357]: I still don't understand... In any case, having this in two commits will improve bisectability which is a clear win in my eyes.

File: core/metacling/src/TClingCallbacks.cxx (hunk 6 above)
[the remark this answers is not in the document]
Comment (hahnjo) [reply core/metacling/src/TClingCallbacks.cxx]: Ok, it seems we need both - I still don't understand why...

File: core/metacling/src/TClingDataMemberInfo.cxx (hunk 3 above)
Comment (hahnjo) [reply core/metacling/src/TClingDataMemberInfo.cxx:357]: FWIW to make the test pass, I need only the changes in `AutoloadLibraryGenerator` and `AutoloadLibraryMU` , not this one. So, why this change and what does it have to do with autoloading?

Comment (vgvassilev) [reply core/metacling/src/TClingDataMemberInfo.cxx:357]: That change was necessary on my machine. In both cases it makes sense that whenever we stop autoloading to disable the callback-based symbol resolution of the JIT as that's what autoloading is intended to limit.

**Comment by dpiparo** [comment]
Hi @vgvassilev , thanks once again for this great development, which was delivered and tested on such a short timescale, basically during the end of the year break. On January 23rd, the last open pre-release of CMSSW_14_0_0 will close: after that date, no ROOT version updates will be possible (see https://github.com/cms-sw/cmssw/issues/43077#issuecomment-1858057605).
Currently, the CMS issue has been worked around, therefore ROOT 6.30 *can* be picked up by CMS. However, it would be much better to have your fix in master and the 6.30 branch for CMS to pick up. That leaves us with a short window of opportunity, which is basically this week. 

Do you think you can factor and succinctly document the changes in the way @hahnjo proposed in order to be able to merge the PR to make your fixes available to CMS? We see LHCb experiencing similar problems - they would also likely benefit from your code.

**Comment by vgvassilev** [comment]
> Hi @vgvassilev , thanks once again for this great development, which was delivered and tested on such a short timescale, basically during the end of the year break. On January 23rd, the last open pre-release of CMSSW_14_0_0 will close: after that date, no ROOT version updates will be possible (see [cms-sw/cmssw#43077 (comment)](https://github.com/cms-sw/cmssw/issues/43077#issuecomment-1858057605)). Currently, the CMS issue has been worked around, therefore ROOT 6.30 _can_ be picked up by CMS. However, it would be much better to have your fix in master and the 6.30 branch for CMS to pick up. That leaves us with a short window of opportunity, which is basically this week.
> 
> Do you think you can factor and succinctly document the changes in the way @hahnjo proposed in order to be able to merge the PR to make your fixes available to CMS? We see LHCb experiencing similar problems - they would also likely benefit from your code.

@dpiparo, I have made the changes that I agree with and responded to the rest of the comments. This PR seems to be ready to go from what concerns me.

File: core/metacling/src/TClingDataMemberInfo.cxx (hunk 3 above)
Comment (hahnjo) [reply core/metacling/src/TClingDataMemberInfo.cxx:357]: I maintain that this change related to initialization of variables and `constexpr` handling. Please note that I'm not commenting on the two checks in `AutoloadLibraryGenerator` and `AutoloadLibraryMU`; while I don't understand them, I acknowledge that they are needed. This change in `core/metacling/src/TClingDataMemberInfo.cxx` does not seem needed.

Comment (vgvassilev) [reply core/metacling/src/TClingDataMemberInfo.cxx:357]: I guess I am lost. This check here avoids going to the expensive `getAddressOfGlobal` operation for `const` and `constexpr` variable declarations. Instead, we evaluate the initializer as we know it is constant and it is faster. The test changes similar to:

```cpp
 gInterpreter->Declare("static const double var4 = 4.2;");
 EXPECT_DOUBLE_EQ(4.2, *(double*)gROOT->GetGlobal("var4")->GetAddress());

 // Make sure ROOT's Core constexpr constants work
 EXPECT_EQ(3000, *(int*)gROOT->GetGlobal("kError")->GetAddress());
```

test that branch here where we get the constant initializer of the variable compute it and return it as a value as required by the interface. In a similar vain:

```cpp
 TGlobal *GeoManagerInfo = gROOT->GetGlobal("gGeoManager");
 TInterpreter::SuspendAutoLoadingRAII autoloadOff(gInterpreter);
 EXPECT_TRUE(-1L == (ptrdiff_t)GeoManagerInfo->GetAddress());
```

tests the fact that when we suspend autoloading we do not resolve `gGeoManager` through the slow path by loading the library and doing dlsym.

Comment (hahnjo) [reply core/metacling/src/TClingDataMemberInfo.cxx:357]: I'm not debating the usefulness of the check or whether it's correct or not. I am complaining that it has *nothing* to do with autoloading, is explained *nowhere* in the commit message nor PR summary, and is *not* required to pass the added test (according to my setup). IMHO this means it *must* be in its own commit with a *proper* explanation, or maybe even in an entirely separate PR.

And I fail to see how your second comment regarding the test involving `gGeoManager` has anything to do with the code I am commenting on in `core/metacling/src/TClingDataMemberInfo.cxx`?

Comment (vgvassilev) [reply core/metacling/src/TClingDataMemberInfo.cxx:357]: In the commit message says: `In addition when computing the offsets of a constant variable
declaration we compute the initializers instead of searching in the shared objects. `

Please suggest wording of the commit message that makes you happy.

> And I fail to see how your second comment regarding the test involving gGeoManager has anything to do with the code I am commenting on in core/metacling/src/TClingDataMemberInfo.cxx?

We've moved:

```cpp
if (Longptr_t addr = reinterpret_cast<Longptr_t (fInterp->getAddressOfGlobal(GlobalDecl(VD))))
         return addr;
```
in terms of priorities - so for me it is quite connected even though it's not on that particular line, it has been downgraded in terms of priority...

Comment (vgvassilev) [reply core/metacling/src/TClingDataMemberInfo.cxx:357]: Okay, I think I responded to that and my reasoning is that’s one feature in my opinion. I fail to see your arguments why I should separate these changes since they are entangled. First we delay calling the slow getSymbolAddress operation and at the same time enable the computation of the constexpr variables from the initializer list.

Comment (hahnjo) [reply core/metacling/src/TClingDataMemberInfo.cxx:357]: To avoid us circling around forever, here is the concrete organization that I have in mind: https://github.com/hahnjo/root/commits/reduce-symbol-search/ As far as I can tell, all tests pass in the intermediate stages. The commit messages are still quite empty, could probably be enhanced. And it's debatable whether the second and third commit should actually be split at this level; but as I expected it seems to be just important that the slow path is only moved after evaluating the initializer of constant variables...

Comment (vgvassilev) [reply core/metacling/src/TClingDataMemberInfo.cxx:357]: Does that sequence works if you revert the workarounds first?

Either way, I don't want to bikeshed more on how to structure commits. I still think these are details that are a matter or preference of the people who do the work. For better or worse ROOT has not have even a policy that tests go with the same commit.

If you feel strongly about that ordering you are welcome to take over the PR. I am not quite sure if that's the way we should review first responders' time-critical fixes, though.

Comment (hahnjo) [reply core/metacling/src/TClingDataMemberInfo.cxx:357]: > Does that sequence works if you revert the workarounds first?

Do you mean moving the last commit first to test the `constexpr`? I can try tomorrow...

> Either way, I don't want to bikeshed more on how to structure commits. I still think these are details that are a matter or preference of the people who do the work. For better or worse ROOT has not have even a policy that tests go with the same commit.

Maybe it's close to bikeshedding and I'm known to be very pedantic regarding commit organization, but in a core critical area such as symbol search, I think every minute we can invest now will potentially save us hours at some point in the future. I pointed out this request from the very beginning, which I think is a very normal code review procedure in functioning open source projects. The response was that it's technically not possible because the changes are entangled which I provided proof that it's not the case. I personally don't see any reason *not* to take the better organization.

> If you feel strongly about that ordering you are welcome to take over the PR. I am not quite sure if that's the way we should review first responders' time-critical fixes, though.

I'm sorry if you feel that way but I will restate that it is my strong personal belief that code review in open source projects should work like this. For me this is not a standard that should be sacrificed in the name of time-criticality (even though that is the reason I tried to follow up as quickly as possible).

Comment (hahnjo) [reply core/metacling/src/TClingDataMemberInfo.cxx:357]: > > Does that sequence works if you revert the workarounds first?
>
> Do you mean moving the last commit first to test the `constexpr`? I can try tomorrow...

Yes, it also works. So you want me to push the re-organized commits to this PR?

**Comment by dpiparo** [comment]
Hi @hahnjo , @vgvassilev : thanks a lot for the exchanges and work on this topic. I was wondering if you can estimate an ETA for the merging of this nice work, key for addressing CMS (and LHCbs) current issues in a sustainable way and for the adoption of ROOT 6.30 by CMS in 2024.

File: core/metacling/src/TClingDataMemberInfo.cxx (hunk 3 above)
Comment (vgvassilev) [reply core/metacling/src/TClingDataMemberInfo.cxx:357]: Sure.

File: core/metacling/src/TCling.cxx (hunk 1 above)
[the remark this answers is not in the document]
Comment (hahnjo) [reply core/metacling/src/TCling.cxx:4729]: It seems I can just remove this and tests still pass on my system. Can you assess if this is still needed in the latest version of the PR? Shall we try reverting it?

**Review by hahnjo — APPROVED** [review]
In any case, the organization is now good from my point of view.

File: core/metacling/src/TCling.cxx (hunk 1 above)
[the remark this answers is not in the document]
Comment (vgvassilev) [reply core/metacling/src/TCling.cxx:4729]: No, it was needed on my system.

**Review by dpiparo — APPROVED** [review]
Dear @vgvassilev , Thanks for these changes. Can you merge them asap? We can then make sure @smuzaffar uses this commit for the CMSSW builds in order for CMS to fully benefit from the improvements.
